File: 1305. All Elements in Two Binary Search Trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {number[]}
 */
var getAllElements = function(root1, root2) {
    const stack1 = [];
    const stack2 = [];
    let current1 = root1;
    let current2 = root2;
    while(true) {
        if (current1) {
            stack1.push(current1);
            current1 = current1.left;
        } else if (current2) {
            stack2.push(current2);
            current2 = current2.left;
        } else if (stack1.length && stack2.length) {
            const node1 = stack1[stack1.length - 1];
            const node2 = stack2[stack2.length - 1];
            if (node1.val > node2.val) {
                result.push(node2.val);
                stack2.pop();
                current2 = node2.right;
            } else {
                result.push(node1.val);
                stack1.pop();
                current1 = node1.right;
            }
        } else if (stack1.length) {
            const node = stack1.pop();
            result.push(node.val);
            current1 = node.right;
        } else if (stack2.length) {
            const node = stack2.pop();
            result.push(node.val);
            current2 = node.right;
        } else {
            break; 
        }
    }
    return result;
};

File: Golang/Tree/1305.all-elements-in-two-binary-search-trees.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func getAllElements(root1 *TreeNode, root2 *TreeNode) []int {
	ret := []int{}
	current1, current2 := root1, root2
	stack1, stack2 := []*TreeNode{}, []*TreeNode{}
	for {
		if current1 != nil {
			stack1 = append(stack1, current1)
			current1 = current1.Left
		} else if current2 != nil {
			stack2 = append(stack2, current2)
			current2 = current2.Right
		} else if len(stack1) > 0 || len(stack2) > 0 {
			if len(stack2) == 0 {
				node := stack1[len(stack1)-1]
				stack1 = stack1[:len(stack1)-1]
				ret = append(ret, node.Val)
				current1 = node.Right
			} else if len(stack1) == 0 {
				node := stack2[len(stack2)-1]
				stack2 = stack2[:len(stack2)-1]
				ret = append(ret, node.Val)
				current2 = node.Right
			} else {
				node1 := stack1[len(stack1)-1]
				node2 := stack2[len(stack2)-1]
				if node1.Val > node2.Val {
					stack2 = stack2[:len(stack2)-1]
					ret = append(ret, node2.Val)
					current2 = node2.Right
				} else {
					stack1 = stack1[:len(stack1)-1]
					ret = append(ret, node1.Val)
					current1 = node1.Right
				}
			}
		} else {
			break
		}
	}
	return ret
}
